v1.8 backports 2020-12-02

nathanjsweet opened the pull request:
v1.8 backports 2020-12-02

 * #14201 -- node: Handle arpinging when remote node is in different L2 (@brb)
 * #14054 -- kpr: ensure DirectRoutingDevice is in devices (@kkourt)
 * #14222 -- vendor: Fix cilium/arping goroutine leak (@jrajahalme)
 * #14238 -- vendor: update etcd to 3.4.14 (@aanm)
 * #14256 -- node: Fix ineffectual assignment (@brb)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 14201 14054 14222 14238 14256; do contrib/backporting/set-labels.py $pr done 1.8; done
```

kkourt committed:
kpr: ensure DirectRoutingDevice is in devices

[ upstream commit 63f19d232b090137dc17fd047b091752b0dd99ba ]

As reported by #14052, there can be a panic if the configured
DirectRoutingDevice is not included in devices. (InitNodePortAddrs() is
called with option.Config.Devices.)

docs state that:
 If the direct routing device does not exist within devices, Cilium
 will add the device to the latter list.

so this patch adds the device to the list.

Fixes #14052

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>
Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

brb committed:
node: Handle case when remote node is in different L2

[ upstream commit af6ca1948d9b2d284f8b94b8a672c4133cab8f2d ]

Previously, insertNeighbor() was assuming that a remote node is in the
same L2 subnet, i.e. directly reachable w/o a gateway. However, this is
not the case for all deployments.

This commit adds a check for detecting whether the remote node is in the
same L2. If it's not, then a gateway IP addr (nexthop) is going to be
arpinged instead of the remote node IP addr.

The missing bit in this commit is a refcounting to avoid redundant arpings
and neigh removals when the gateway is used to access more than one
remote node.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

brb committed:
node: Refcount neighbour entries

[ upstream commit b78b3b7f259d842a96142c7c900d4b293fc7d23b ]

To avoid redundant pings and neigh entry removals when the entry is
still used by other node (happens when two or more nodes can be accessed
through the same gateway).

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

jrajahalme committed:
vendor: Fix cilium/arping goroutine leak

[ upstream commit 24d44500e40af599dfc1b932be0dac1b75504889 ]

This fixes a privileged runtime test failure caused by leaked goroutines on arpings with no response.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

aanm committed:
vendor: update etcd to 3.4.14

[ upstream commit 7b0037c31f8b57940f395c765306db158b15cd1a ]

This version fixes etcd error "auth token invalid after
watch reconnects" https://github.com/etcd-io/etcd/issues/11954.

Signed-off-by: André Martins <andre@cilium.io>

kkourt reviewed (approved):
My patch looks good.

brb committed:
node: Fix ineffectual assignment

[ upstream commit 354d42bfa2eedce150c22820fdc366e1f3b948f9 ]

/home/travis/gopath/src/github.com/cilium/cilium/pkg/datapath/linux/node_linux_test.go:1096:14:
ineffectual assignment to err

Reported-by: André Martins <andre@cilium.io>
Fix: b78b3b7f25 ("node: Refcount neighbour entries")
Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
